Correct home yaw when a valid heading is acquired

fiam opened the pull request:
This is specially relevant in FW without compass, since the home
location will typically be stored while the craft hasn't yet moved
and a CoG derived heading is not determined yet.

The navigation code checks when the heading transitions from invalid
to valid in updateActualHeading(), calculates the rotation from the old
invalid heading (using bootup heading as 0º) to the real one and
applies to posControl.homePosition.yaw.

In order to save a bit of memory, isHomeValid has been replaced
with a bitmask which stored several flags relative to the home
location.

fiam committed:
Correct home yaw when a valid heading is acquired

This is specially relevant in FW without compass, since the home
location will typically be stored while the craft hasn't yet moved
and a CoG derived heading is not determined yet.

The navigation code checks when the heading transitions from invalid
to valid in updateActualHeading(), calculates the rotation from the old
invalid heading (using bootup heading as 0º) to the real one and
applies to posControl.homePosition.yaw.

In order to save a bit of memory, isHomeValid has been replaced
with a bitmask which stored several flags relative to the home
location.

digitalentity commented on src/main/navigation/navigation.c:
```diff
@@ -1554,9 +1554,26 @@ void updateActualAltitudeAndClimbRate(bool estimateValid, float newAltitude, flo
+        (posControl.homeFlags & NAV_HOME_VALID_HEADING) == 0) {
+
+        int32_t yawRotation = newHeading - posControl.actualState.yaw;
+        posControl.homePosition.yaw += yawRotation;
```
Shouldn't we simply rewrite `posControl.homePosition.yaw` here? Why the `yawRotation` magic?

fiam replied:
You need the rotation because you might arm, turn and then launch. If you turned 90º before launching the "fake" yaw at that point will be 90º and home will be at 0º (assuming you armed without rotating since battery was plugged in). Now you take off and once you get the real heading yaw is becoming 180º. That means your fake yaw was 90º off to the left, hence your home yaw needs that same correction since it was calculated with the fake one.

The problem with storing the first acquired yaw as home yaw is that you might unintentionally throw the plane a bit off from the direction you intended or even the pitch/roll stabilisation might send it a bit to one side (launch mode doesn't try to keep heading, but level). Using the yaw when you arm allows more predictable values for the pilot.

digitalentity replied:
Ok, now I understand. Still, this code looks somewhat "hacky" to me, but I can't think about a way of improving it.

fiam replied:
Maybe we can rename `yawRotation` to `fakeToRealYawOffset`?

digitalentity replied:
Agreed, it will make things easier to understand. A comment, explaining the code, will also be quite helpful.

digitalentity commented on src/main/navigation/navigation.c:
```diff
@@ -1698,24 +1715,29 @@ void setHomePosition(const fpVector3_t * pos, int32_t yaw, navSetWaypointFlags_t
     if ((useMask & NAV_POS_UPDATE_HEADING) != 0) {
         // Heading
         posControl.homePosition.yaw = yaw;
+        if (posControl.flags.estHeadingStatus >= EST_USABLE) {
```
`setHomePosition` may be called with arbitrary `yaw`. We shouldn't check for `estHeadingStatus` in this function. We could check for valid heading before calling `setHomePosition` though.

fiam replied:
Opps. Nice catch! Thing is, for the best result we actually want to store it with the "fake" yaw, then correct it, since we don't don't have a valid yaw at arming time in crafts without mag. Maybe we can pass the `homeFlags` to `setHomePosition()`?

digitalentity replied:
Not sure, need to think about it. `setHomePosition` has access to homeFlags anyway

fiam replied:
I mean we can let the caller tell `setHomePosition()` which validity flags it should set for the home position.

digitalentity replied:
Ah, that's definitely a good idea

fiam committed:
Make setHomePosition() accept a new argument indicating the home flags

Callers can specify wether XY, Z and HEADING are valid for the new
home position.

fiam committed:
Rename yawRotation to fakeToRealYawOffset

Also, add a small comment explaining the logic behind the home
yaw adjustment.
